From bcb5120535f09df406d187ce8b367d31bbeef0a7 Mon Sep 17 00:00:00 2001
From: Gary Lin <glin@suse.com>
Date: Thu, 12 Oct 2017 17:24:04 +0800
Subject: [PATCH 28/55] shim: Show the warning for the CA check result

After verifying the image, a warning will show if the less strict CA
check is used.

Signed-off-by: Gary Lin <glin@suse.com>
---
 shim.c | 30 ++++++++++++++++++++++++++++++
 1 file changed, 30 insertions(+)

diff --git a/shim.c b/shim.c
index 6638a26..f29f392 100644
--- a/shim.c
+++ b/shim.c
@@ -406,6 +406,24 @@ static BOOLEAN verify_eku(UINT8 *Cert, UINTN CertSize)
 	return TRUE;
 }
 
+static void show_ca_warning()
+{
+	CHAR16 *text[9];
+
+	text[0] = L"WARNING!";
+	text[1] = L"";
+	text[2] = L"The CA certificate used to verify this image doesn't   ";
+	text[3] = L"contain the CA flag in Basic Constraints or KeyCertSign";
+	text[4] = L"in KeyUsage. Such CA certificates will not be supported";
+	text[5] = L"in the future.                                         ";
+	text[6] = L"";
+	text[7] = L"Please contact the issuer to update the certificate.   ";
+	text[8] = NULL;
+
+	console_reset();
+	console_print_box(text, -1);
+}
+
 static CHECK_STATUS check_db_cert_in_ram(EFI_SIGNATURE_LIST *CertList,
 					 UINTN dbsize,
 					 WIN_CERTIFICATE_EFI_PKCS *data,
@@ -422,12 +440,16 @@ static CHECK_STATUS check_db_cert_in_ram(EFI_SIGNATURE_LIST *CertList,
 			CertSize = CertList->SignatureSize - sizeof(EFI_GUID);
 			if (verify_x509(Cert->SignatureData, CertSize)) {
 				if (verify_eku(Cert->SignatureData, CertSize)) {
+					clear_ca_warning();
 					IsFound = AuthenticodeVerify (data->CertData,
 								      data->Hdr.dwLength - sizeof(data->Hdr),
 								      Cert->SignatureData,
 								      CertSize,
 								      hash, SHA256_DIGEST_SIZE);
 					if (IsFound) {
+						if (get_ca_warning()) {
+							show_ca_warning();
+						}
 						tpm_measure_variable(dbname, guid, CertSize, Cert->SignatureData);
 						drain_openssl_errors();
 						return DATA_FOUND;
@@ -1049,11 +1071,15 @@ static EFI_STATUS verify_buffer (char *data, int datasize,
 		/*
 		 * Check against the shim build key
 		 */
+		clear_ca_warning();
 		if (sizeof(shim_cert) &&
 		    AuthenticodeVerify(cert->CertData,
 			       cert->Hdr.dwLength - sizeof(cert->Hdr),
 			       shim_cert, sizeof(shim_cert), sha256hash,
 			       SHA256_DIGEST_SIZE)) {
+			if (get_ca_warning()) {
+				show_ca_warning();
+			}
 			update_verification_method(VERIFIED_BY_CERT);
 			tpm_measure_variable(L"Shim", SHIM_LOCK_GUID,
 					     sizeof(shim_cert), shim_cert);
@@ -1068,11 +1094,15 @@ static EFI_STATUS verify_buffer (char *data, int datasize,
 		/*
 		 * And finally, check against shim's built-in key
 		 */
+		clear_ca_warning();
 		if (vendor_cert_size &&
 		    AuthenticodeVerify(cert->CertData,
 				       cert->Hdr.dwLength - sizeof(cert->Hdr),
 				       vendor_cert, vendor_cert_size,
 				       sha256hash, SHA256_DIGEST_SIZE)) {
+			if (get_ca_warning()) {
+				show_ca_warning();
+			}
 			update_verification_method(VERIFIED_BY_CERT);
 			tpm_measure_variable(L"Shim", SHIM_LOCK_GUID,
 					     vendor_cert_size, vendor_cert);
-- 
2.17.1

